### DISCUSSION TITLE
Using self-hosted Directus as a service in a client project

### DISCUSSION THREAD

**Discussion by Prakhar225** [directus/directus#15688] [Archive]
I am working on a project for a client and want to use Directus as an API server. There are no changes to the Directus itself and all the communication is through HTTP requests and webhooks. Does that mean the entire project has to be under GPL?
I was pretty sure that was not the case but this left me pretty confused https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation

**Comment by ctholho** [comment]
I'm not from the team and I'm not a lawyer. ~~That being said, as I understand it, you have to offer your client the possibility to get the source of Directus, that's it.~~

edit: **easiest:** You release your project to your client as GPL. That means they (and only they) get the source code from you. They still could obtain exclusive usage rights, if you wish/can grant them. They're just not allowed to resell your software themselves without it being GPL.
**Still ok:** Offer your client the possibility to get the source of Directus, that's it. If you (or someone else) installs your application for your client on a server they're renting and you use a docker-compose or other config files you're not distributing the "binaries" yourself, so don't use GPL.

I'm just researching this myself, so take it with a grain of salt – but I'm looking into doing the same as you and I was worried that my client won't hire my services and avoid using Directus because they have to release their source code to the public (or say to the app users themselves).

Well, it turns out that's not the case.

edit: See this very short and to the point answer by one of the maintainers. #14422. Now, on to the rest of my lengthy answer.

This part of the [GPL FAQ](https://www.gnu.org/licenses/gpl-faq.html#UnreleasedMods) makes clear what a user is in our case.
>*A company is running a modified version of a GPL'ed program on a web site. Does the GPL say they must release their modified sources?*
>The GPL permits anyone to make a modified version and use it without ever distributing it to others. What this company is doing is a special case of that. Therefore, the company does not have to release the modified sources.

The user is your client in that case. If you distribute it only to your client, you are only forced to give your client the source code of Directus.

In regards to your aggregation confusion, see [this great answer in regards to containerization](https://opensource.stackexchange.com/a/8393/28172):
>As hinted at above, the A/GPLv3 allows you to distribute A/GPLv3 covered software alongside other software, without the A/GPLv3 affecting the license of this software. They are completely separate softwares, they just happen to be stored on the same filesystem.

Tangentially:
If you're configuring webhooks or Directus flows, those are data in your DB not part of Directus. From your link, a bit further up it's stated:
> Thus, for instance, you must make the **source code** [my emphasis] available to the users of the program as described in the GPL, and they [your customer] must be allowed to redistribute and modify it as described in the GPL.

They explicitly state 'source code'. The GPL excludes configuration details that have only influence on the runtime. That means your customer is not allowed to just use your configurations and redistribute them, if that was your concern.

Also, if you're building a website that connects with Directus and you use the directus-sdk (licensed under MIT, btw) your website doesn't need to be GPL.

I'm still trying to figure out what that means for extensions. The extension-sdk is licensed as GPL but as a build tool it doesn't force you to use GPL for your extensions. But it's very strictly linked to directus. The main thing is that extensions (e.g. endpoints) use expressJS and are therefore so generally applicable that you might get away with it.

You can build a SaaS with it and keep every modification of Directus for yourself. You're just not allowed to offer your "enterprise" customers an on-premise build Docker image of your Directus derivate and declare your release as MIT.

You could still sell your custom made docker-compose YAML and license them under any license you wish.

**Comment by rijkvanzanten** [comment]
Thanks for the in depth train of thought @ctholho 🙂 

From my perspective, the answer to:

> There are no changes to the Directus itself and all the communication is through HTTP requests and webhooks. Does that mean the entire project has to be under GPL?

Is *no*. To me (as far as I understand it, and obviously IANAL), _usage_ of the (unmodified) distributed version of Directus isn't bound by the source code license. Only when you'd make a fork of Directus itself you would have to open source your fork of Directus under the same license.

**Reply by rijkvanzanten** [reply]
As for extensions, it's a bit of a legal gray area, but as far as I understand it we're not statically linking any parts of the Directus source in extensions[^1], so it's not bound by the source code license. 

[^1]: In the extensions, Directus provides parameters to a handler function, but you don't have to import / bundle any parts of Directus' source code directly in order to build extensions. From what I understand, that means it's not bound by the source code license

**Reply by ctholho** [reply]
Thanks @rijkvanzanten.
Although I'm very tempted to take your word for it, I'm afraid it's not that simple.

The GPL FAQ says this [about extensions and plugins](https://www.gnu.org/licenses/gpl-faq.en.html#GPLPlugins):
> If the main program dynamically links plug-ins, and they make function calls to each other and share data structures, we believe they form a single combined program

I believe e.g. a custom interface does satisfy that requirement

It goes on to state:
> If the main program dynamically links plug-ins, but the communication between them is limited to invoking the ‘main’ function of the plug-in with some options and waiting for it to return, that is a borderline case.

That would be the case for simple custom endpoints.

**Reply by rijkvanzanten** [reply]
> [...] I'm afraid it's not that simple.

Oh certainly not! My comments above are 100% _To me, in my opinion, as far as I understand it, and obviously IANAL_ replies. 

Based on that second statement "limited to invoking the 'main' function", I _believe_ all of our extension types would fall in that bucket. App extensions are also invoked from a single main "defineX" function call similar to API extensions 🤔
